learn/appendix_opengl/KHR_debug.md

```diff
@@ -0,0 +1,194 @@
+
+### `GL.DebugMessageCallback()`
+This function allows you to set the callback that will receive debug messages.
+It is only mentioned here as it has been extensively described in the
+[Debug Callback](debug_callback.md) article.
+
+### `GL.DebugMessageControl()`
+This function allows you to control the debugging messages that will be received
+by the callback.
+```cs
+void GL.DebugMessageControl(
+    DebugSourceControl source,      /* Select debug sources or all. */
+    DebugTypeControl type,          /* Select message types or all. */
+    DebugSeverityControl severity,  /* Select message severity or all */
+    int numIds,     /* Number of IDs to supply. */
+    int[] ids,      /* Array of IDs */
+    bool isEnabled  /* Enable or disable. */
+);
+```
+The parameter `source` is the message source filter. It can be your own
+application or components of the OpenGL or windowing API. `type` filters which
+message types to control. `severity` filters which severity to control. In
+addition to this, the enum value `.DontCare` does not filter any message type.
+When `numIds` is positive, ids is an array of message IDs to filter. You can use
+`0` and `System.Array.Empty<int>()` if you don't want to feed the function any
+message IDs. And finally `isEnabled` controls if the filtered messages are to be
+permitted or ignored.
+
+This function is especially useful for turning down warnings that you do not
+desire to see. Since the messages generated are driver specific, it might be
+useful to keep a list of IDs you might want to mute or permit per vendor, that
+way your debugging logs are clear of any frequent informational messages.
+
+### `GL.DebugMessageInsert()`
+The following function allows you to insert a custom message into the debug
+message stream. The parameters are similar to that of `GL.DebugMessageControl`
+filter parameters.
+```cs
+void GL.DebugMessageInsert(
+    DebugSourceExternal source, /* Source of the message. */
+    DebugType type,             /* Type of the message. */
+    int id,                     /* An ID to associate with this message. */
+    DebugSeverity severity,     /* Severity of the message. */
+    int length,                 /* Length of the message or -1 for automatic. */
+    string message              /* The message to insert. */
+);
+```
+
+### `GL.GetDebugMessageLog()`
+In the abscense of a debug message callback, this function can be used to
+retreive debug messages that have been enqueued by the OpenGL context.
+```cs
+int GL.GetDebugMessageLog(
+    int number,                 /* Number of messages to retreive. */
+    int bufferSize,             /* Message buffer size. */
+    out DebugSource sources,    /* Retreived message sources. */
+    out DebugType types,        /* Retreived message types. */
+    out int ids,                /* Retreived message IDs. */
+    out DebugSeverity severity, /* Retreived message severities */
+    out int length,             /* Retreived message lengths. */
```

Comment:
This parameter is called `lengths` in the refpages, but I understand the singular, maybe we want to change the rest of the plurals for the `out` overload.